compiler/testData/codegen/box/assert/jvm/initializerAssertionsEnabled.kt

```diff
@@ -0,0 +1,53 @@
+// TARGET_BACKEND: JVM
```

Comment:
Could you also add 'initializerAssertionsDisable.kt' test by 'interfaceAssertionsDisabled.kt/interfaceAssertionsEnabled.kt' analogue?

Actually compiler also moves all initializer from class companion to outer class. And here is an interesting corner case with assertions. @ilmirus Seems this  code is not supported in current backend:
```
open class Bar {    
    companion object {
        val barAssertionThrown = try {
            assert(false) 
            false
        } catch(error: java.lang.AssertionError) {
            true
        }
    }
}
...
I don't see any $assertionsDisabled neither in companion nor in Bar. And if understand correctly this field initialization should  be performed only in companion clinit (that require some trick at least in IR backend cause all initialization is poped up in outer class)

Reply:
@sfs NB:  corner case support is not necessery for this PR

Reply:
This code is, in fact, supported in current backend. The `$assertionsDisabled` field is generated in `Bar` class and not in companion. This, it seems, is in line with intended behavior for inner classes: if `assert` is used in inner class, we shall generate `$assertionsDisabled` in outer class.

Reply:
However, in cases of nested and inner classes the old backend generates the field in inner class. This is known bug.

Reply:
For comparison: Javac generates a field in an inner class, but initializes the assertion status based on the outer class. In either case, there's no way to selectively enable/disable assertions in an inner class.

I will change the code to generate the field in the outer class and add some tests for this. We will still need to add support for this in `IrInlineCodegen.generateAssertFieldIfNeeded`, though.

Reply:
I've added 5 more tests for setting the assertion status per class: for toplevel, companion, local, nested, and inner classes. Currently, local, nested, and inner classes do not work in the JVM backend.

At the same time I've fixed these cases in the JVM_IR backend.

Reply:
@sfs Thanks, I will merge PR as is after rr. 

@ilmirus Actually as mentioned by @sfs javac generates field in each nested class but initializes it with outer class state:
```
    LDC LOuter;.class
    INVOKEVIRTUAL java/lang/Class.desiredAssertionStatus ()Z
    PUTSTATIC Inner.$assertionsDisabled : Z
```
As I remember `LDC Class` doesn't trigger class initialization

Reply:
@sfs last request could you also add muted bytecode test with proper `$assertionsDisabled` initialization though 'desiredAssertionStatus(outer)' ?

Reply:
> As I remember `LDC Class` doesn't trigger class initialization

That's true ([jvm spec](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-5.html#jvms-5.5)) and explains the Java behavior. For compatibility we should do the same in Kotlin, but this is not a backwards compatible change. For example, consider the following code.

```kotlin
class Outer {
  companion object {
    init { error("") }
  }

  class Inner {
    init { assert(true) }
  }
}

fun main() {
  Outer.Inner()
}
```

When compiled with `-Xassertions=jvm`, `main` will fail with an exception, since we trigger class initialization for `Outer` (through `getstatic Outer.$assertionsDisabled` in `Outer$Inner.<init>`). The equivalent code in Java --- or for that matter the Kotlin code with any other assertions mode --- will run just fine.

Do you know in which language version `-Xassertions=jvm` was introduced? Is the feature still experimental? Otherwise I can file a language committee issue for this.

> could you also add muted bytecode test with proper nested class $assertionsDisabled initialization though 'desiredAssertionStatus(outer)' ?

Will do! For the reasons above I don't want to codify the current behavior, but I can at least add a test which checks that we only look at the assertion status in the Outer class.

Reply:
Correction: the code example above doesn't currently work in the JVM backend anyway (since the assertion status is initialized based on the inner class). Did it work in a previous version?

Reply:
@max-kammerer I've added a simple bytecode test in `compiler/testData/codegen/bytecodeText/assert/jvmNestedClass.kt`. It's not very precise and I would like to expand on this once it's clear what we should do about the issue above.

Reply:
All modes of `-Xassertions` are experimental, so no language committee approval is needed, we're free to change behavior as we please here. From my (slightly out-of-context) view, it'd be ideal to generate bytecode similar to javac both in the old and in the IR backend, but we can basically postpone this until the feature is graduated (which I think is not yet planned for `-Xassertions` in 1.4).

Reply:
@sfs Yes, this feature is still experimental and could be changed without compatibility respect.  There is a lack of proper nested/local class support in current backend. The test looks good, please just add 
'
// 0 PUTSTATIC Outer
// 1 PUTSTATIC Outer$Inner.\$assertionsDisabled
'
 to be sure that field is placed in correct class

Reply:
> All modes of `-Xassertions` are experimental, so no language committee approval is needed, we're free to change behavior as we please here.

> Yes, this feature is still experimental and could be changed without compatibility respect.

That's perfect, in this case I'll just change the lowering to mimic the javac behavior and add more tests.

Reply:
@max-kammerer I've changed the lowering as discussed, made the bytecode text test more precise, and added a new box test to ensure that we don't run class initializers just because we use assertions in a nested class. Except for inlining, everything should work as in javac now.

Incidentally, the bytecode text test for asserts in crossinline lambda arguments to inline functions (`compiler/testData/codegen/bytecodeText/assert/jvmCrossinlineAssertInLambda.kt`) doesn't seem correct to me. The test checks that the assertion status is set based on the generated lambda class (`LDC LA\$inlineSite\$\$inlined\$inlineMe\$1;.class`), instead of based on the top level class.

Reply:
Yes, it's wrong test

Reply:
Ok, I've fixed the test as well.